# Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

## Instructions:

This challenge consists of the following subsections:

* Split the Data into Training and Testing Sets

* Create a Logistic Regression Model with the Original Data

* Predict a Logistic Regression Model with Resampled Training Data 

### Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

1. Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

2. Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

    > **Note** A value of `0` in the “loan_status” column means that the loan is healthy. A value of `1` means that the loan has a high risk of defaulting.  

3. Check the balance of the labels variable (`y`) by using the `value_counts` function.

4. Split the data into training and testing datasets by using `train_test_split`.

### Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

1. Fit a logistic regression model by using the training data (`X_train` and `y_train`).

2. Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.

4. Answer the following question: How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use `RandomOverSampler`.

To do so, complete the following steps:

1. Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

2. Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

3. Evaluate the model’s performance by doing the following:

    * Calculate the accuracy score of the model.

    * Generate a confusion matrix.

    * Print the classification report.
    
4. Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

### Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the `README.md` file included in your GitHub repository.

Structure your report by using the report template that `Starter_Code.zip` includes, and make sure that it contains the following:

1. An overview of the analysis: Explain the purpose of this analysis.


2. The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

3. A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:

# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending = pd.read_csv(Path('.//Resources/lending_data.csv'))

# Review the DataFrame
df_lending.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending['loan_status']

# Separate the X variable, the features
X = df_lending.drop(columns='loan_status')

In [4]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
data_org = y.value_counts()
print(data_org)

0    75036
1     2500
Name: loan_status, dtype: int64


In [7]:
data_balance_org = data_org[1] / data_org[0]
unhealthy_loans = data_balance_org
healthy_loans = 1 - data_balance_org

print(f'Dataset displays the following data imbalance: data with status 0 is {healthy_loans:.2f} %, where data with status 1 is {unhealthy_loans:.2f}')

Dataset displays the following data imbalance: data with status 0 is 0.97 %, where data with status 1 is 0.03


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_original_model = model.fit(X_train,y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
y_original_pred = lr_original_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
bas_org = balanced_accuracy_score(y_test,y_original_pred)

print(round(bas_org,5))

0.95205


In [12]:
# Generate a confusion matrix for the model
confusion_matrix(y_test,y_original_pred)

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [13]:
# Print the classification report for the model
class_rep_org = classification_report_imbalanced(y_test, y_original_pred)
print(class_rep_org)

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.91      1.00      0.95      0.91     18765
          1       0.85      0.91      0.99      0.88      0.95      0.90       619

avg / total       0.99      0.99      0.91      0.99      0.95      0.91     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 

**1. Dataset Balance**
* IMBALANCED: There were many more 'healthy loans' 97% in the data set than 'high-risk' loans 3%. (75036 healthy loans vs 2500 high-risk loans)

**2. Accuracy**
* VERY GOOD: Approximately 95.2% of the loans in the test data were accurately categorized by the model. But the model could have had high accuracy by simply predicting all healthy loans correctly due to the dataset imbalance.

**3. Precision:**
* VERY GOOD / GOOD: For 0 is 100% for 1 is 85% which is still good but might not be good enough for the purpose of this model, detecting high risk loans. 

* Healthy = 18663 / (18663 + 56) = 99.7%
* Unhealthy = 563 / (563 + 102) = 84.7%

**4. Recall**
* VERY GOOD / GOOD: For 0 is 99%, for 1 is 91% which looks good but might not be good enough for the purpose of this model. Model classified correclty 91% of the high risk loans, but still 56 loans were not correctly identified as high risk. 

* Healthy = (18663/(102+18663) = 99.4%
* Unhealthy = (563/(56+563) = 91%

**5. F1-Score**
* VERY GOOD/ GOOD: For 0 is 100%, for 1 is 88%. F1 score (the harmonic mean) is a single summary statistic for the precision and the recall. The model might have a high F1 score. But, that can prove deceptive, because the model might do a good job of predicting only our healthy loans. However, we’re interested in the unhealthy loans.

* F1 = 2 * (precision * recall) / (precision + recall)
* Healthy = 2 *((99.7% * 99.4%) / (99.7% + 99.4%)) = 99.7%
* Unhealthy = 2 * ((84.7% * 91%) / (84.7% + 91%)) = 87.7%


## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [20]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_oversampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = random_oversampler.fit_resample(X_train,y_train)

In [24]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [25]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
lr_resampled_model= model.fit(X_resampled,y_resampled)

# Make a prediction using the testing data
y_resampled_pred = lr_resampled_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [26]:
# Print the balanced_accuracy score of the model 
# Print the balanced_accuracy score of the model
bas_resamp = balanced_accuracy_score(y_test,y_resampled_pred)

print(round(bas_resamp,5))

0.99368


In [27]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_resampled_pred)

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [28]:
# Print the classification report for the model
print(classification_report_imbalanced(y_test, y_resampled_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       1.00      0.99      0.99      1.00      0.99      0.99     18765
          1       0.84      0.99      0.99      0.91      0.99      0.99       619

avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
**1. Dataset Balance**
* BALANCED: There are as many 'healthy loans' as 'high-risk' loans in the dataset, this is achieved by RandomOverSampler Module.

**2. Accuracy**
* VERY GOOD: Approximately 99.37% of the loans in the resampled compared to 95.2% in the original data. 

**3. Precision:**
* VERY GOOD / GOOD: For 0 is 100% for 1 is 84% which is slightly lower than in the original data. 

* Healthy = 18649 / (18649 + 4) = 99.98%
* Unhealthy = 615 / (615 + 116) = 84.1%

**4. Recall**
* VERY GOOD / VERY GOOD: For 0 is 99%, for 1 is 99%. Model with resampled data classified correclty 91% of the high risk loans, only 4 loans were not correctly identified as high risk. 

* Healthy = (18649/(116+18649) = 99.38%
* Unhealthy = (615/(4+615) = 99.35%

**5. F1-Score**
* VERY GOOD/ GOOD: For 0 is 100%, for 1 is 91%. Which is much higher than the approximate 88% based on the original data set. 

* F1 = 2 * (precision * recall) / (precision + recall)
* Healthy = 2 *((99.98% * 99.38%) / (99.98% + 99.37%)) = 99.68%
* Unhealthy = 2 * ((84.1% * 99.35%) / (84.1% + 99.35%)) = 91.09%

All in, the model using resampled data was better at detecting unhealthy loans than what the model generated using the original, imbalanced dataset.